core/primitives/src/types.rs

```diff
@@ -27,6 +27,23 @@ pub type Balance = u128;
 /// Gas is a type for storing amount of gas.
 pub type Gas = u64;
 
+/// Number (or amount) of blocks (or block heights) in current context.
+// TODO discuss
+// Technically, blocks size is not equal to heights difference.
+// For example, 50 blocks don't imply heights from H to (H + 50).
+// It may lead to incorrect assumptions in future.
+// We can replace NumBlocks with HeightHorizon or something better.
+pub type NumBlocks = u64;
+/// Number of shards in current group.
+pub type NumShards = u64;
+/// Number of validators in current group.
+pub type NumValidators = u64;
+/// Number of block producers in current group.
+pub type NumBlockProducers = u64;
+/// Number of seats of block producers in current group.
+// TODO MOO #1855: understand the difference between NumValidators, NumBlockProducers and NumSeats clearly
+pub type NumSeats = u64;
```

Comment:
I completely support your vision about having explicit types instead of reusing some types just because they happen to have enough bytes to store the information!

However, I want to ask if we want to make it even more strict. `type` is just making an alias, and thus, one still can mess up and do `NumBlocks + NumShards - NumValidators`. Should we implement these types as single-value tuples with only a few explicit operations implemented?

Reply:
Personally, I'd prefer to keep it as agreement, not as strict rule.
Normally types should work as helpers for understanding the nature of the value. It should be okay to calculate `NumValidators + NumShards` for some specific reasons - those ones always appear.
I'm afraid that making types strict may lead to the following:
1. More casts, more methods; more code for compiler, not for developer.
2. "Temporary workarounds" to avoid using strict typing.
3. Difficulties with special cases.
4. Huge refactoring (once) which takes time and attention of code owners.

It will never too late to change to strict types. We can start with agreement easily and then see how does it work.